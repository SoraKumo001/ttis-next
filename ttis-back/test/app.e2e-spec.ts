import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { AppModule } from './../src/app.module';
import fetch from 'isomorphic-unfetch';
import { setContext } from 'apollo-link-context';

import {
  ApolloClient,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject,
  gql,
  ApolloLink,
} from 'apollo-boost';

interface AuthLink extends ApolloLink {
  setToken: (token: string) => void;
}
const createAuthLink = (options: HttpLink.Options) => {
  let bearerToken: string;
  const apolloLink = setContext((_, { headers }) => ({
    headers: { ...headers, authorization: `bearer ${bearerToken || ''}` },
  })).concat(new HttpLink(options)) as AuthLink;
  apolloLink.setToken = (token) => {
    bearerToken = token;
  };
  return apolloLink;
};

describe('GraphQL', () => {
  let app: INestApplication;
  let client: ApolloClient<NormalizedCacheObject>;
  let link: AuthLink;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();
    app = moduleFixture.createNestApplication();
    await app.init();
    await app.listen(0);
    link = createAuthLink({
      fetch,
      uri: `http://localhost:${app.getHttpServer().address().port}/graphql`,
    });
    client = new ApolloClient({
      link,
      cache: new InMemoryCache().restore({}),
    });
  });
  const QUERY_LOGIN = gql`
    mutation Login($name: String!, $password: String!) {
      login(name: $name, password: $password) {
        token
        user {
          name
          info
        }
      }
    }
  `;
  const QUERY_CURRENT_USER = gql`
    query CurrentUser {
      currentUser {
        id
        name
        info
      }
    }
  `;
  it('mutation login', async () => {
    let result = await client.mutate({
      mutation: QUERY_LOGIN,
      variables: { name: 'admin', password: '' },
    });
    expect(result).toMatchSnapshot();
    link.setToken(result.data.login.token);
    result = await client.query({ query: QUERY_CURRENT_USER });
    expect(result).toMatchSnapshot();
  });
});
