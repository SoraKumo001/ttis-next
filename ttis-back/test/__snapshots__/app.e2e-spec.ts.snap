// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL mutation login 1`] = `
Object {
  "data": Object {
    "login": Object {
      "__typename": "Login",
      "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwicGFzc3dvcmQiOiJlM2IwYzQ0Mjk4ZmMxYzE0OWFmYmY0Yzg5OTZmYjkyNDI3YWU0MWU0NjQ5YjkzNGNhNDk1OTkxYjc4NTJiODU1IiwiaWF0IjoxNTg3NjAzNTc2LCJleHAiOjE1ODc2MDk1NzZ9.xqHM4PzUXbJgohVjgfHWVzgum6Jd66pLzfWbNUD8QVA",
      "user": Object {
        "__typename": "User",
        "info": "{}",
        "name": "admin",
      },
    },
  },
}
`;

exports[`GraphQL mutation login 2`] = `
Object {
  "data": Object {
    "currentUser": Object {
      "__typename": "User",
      "id": 1,
      "info": "{}",
      "name": "admin",
    },
  },
  "loading": false,
  "networkStatus": 7,
  "stale": false,
}
`;
